src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader.cs

```diff
@@ -204,29 +204,26 @@ private void ResetReader()
         /// <summary>
         /// Get the next segment with available space, if any.
         /// </summary>
-        [MethodImpl(MethodImplOptions.NoInlining)]
         private void GetNextSpan()
         {
-            if (!Sequence.IsSingleSegment)
+            CurrentSpanIndex = 0;
+            CurrentSpan = default;
+
+            SequencePosition previousNextPosition = _nextPosition;
+            while (_nextPosition.GetObject() != null && Sequence.TryGet(ref _nextPosition, out ReadOnlyMemory<T> memory, advance: true))
             {
-                SequencePosition previousNextPosition = _nextPosition;
-                while (Sequence.TryGet(ref _nextPosition, out ReadOnlyMemory<T> memory, advance: true))
+                _currentPosition = previousNextPosition;
+                if (memory.Length > 0)
+                {
+                    CurrentSpan = memory.Span;
+                    return;
```

Comment:
As a micro-optimization, consider avoiding a return within a loop and also having a single return point by using a goto.